File: solutions/python/18.py
from re import compile, search, sub, IGNORECASE


def find_in_agenda(agenda: str, phone: str) -> dict | None:
    phone_regex_string = "\\+\\d{1,2}-\\d{3}-\\d{3}-\\d{3}"
    name_regex_string = "<[\\w\\s]+>"
    phone_regex = compile(phone_regex_string, IGNORECASE)
    name_regex = compile(name_regex_string, IGNORECASE)
    contacts = agenda.split("\n")

    def match_contact(contact: str) -> bool:
        return search(phone_regex, contact)[0].find(phone) != -1

    matched_contacts = filter(match_contact, contacts)

    def contact_info(contact: str) -> dict:
        return {
            "name": search(name_regex, contact)[0][1:-1],
            "address": sub(phone_regex, "", sub(name_regex, "", contact)).strip()
        }

    matched_contacts_info = list(map(contact_info, matched_contacts))
    if (len(matched_contacts_info) == 1):
        return matched_contacts_info[0]
    else:
        return None


if __name__ == "__main__":
    agenda = """+34-600-123-456 Calle Gran Via 12 <Juan Perez>
Plaza Mayor 45 Madrid 28013 <Maria Gomez> +34-600-987-654
<Carlos Ruiz> +1-800-555-0199 Fifth Ave New York"""

    agenda1 = find_in_agenda(agenda, "34-600-123-456")
    print(agenda1)
    # { name: "Juan Perez", address: "Calle Gran Via 12" }

    agenda2 = find_in_agenda(agenda, "600-987")
    print(agenda2)
    # { name: "Maria Gomez", address: "Plaza Mayor 45 Madrid 28013" }

    agenda3 = find_in_agenda(agenda, "111")
    print(agenda3)
    # null
    # Explanation: No results

    agenda4 = find_in_agenda(agenda, "1")
    print(agenda4)
    # null
    # Explanation: Too many results

"""
  Score: ★★★★★
"""

File: solutions/ts/18.ts
function findInAgenda(
  agenda: string,
  phone: string
): { name: string; address: string } | null {
  const phoneRegexString = "\\+\\d{1,2}(-\\d{3}){3}";
  const nameRegexString = "<[\\w\\s]+>";
  const phoneRegex = new RegExp(phoneRegexString, "gi");
  const nameRegex = new RegExp(nameRegexString, "gi");
  const removeRegex = new RegExp(
    `(${phoneRegexString})|(${nameRegexString})`,
    "gi"
  );
  const contacts = agenda.split("\n");
  const contactsInfo = contacts
    .filter((contact) => {
      const number = contact.match(phoneRegex)![0];
      return number.includes(phone);
    })
    .map((contact) => {
      const name = contact.match(nameRegex)![0].slice(1, -1);
      const address = contact.replaceAll(removeRegex, "").trim();
      return { name, address };
    });
  return contactsInfo.length === 1 ? contactsInfo[0] : null;
}

const agenda = `+34-600-123-456 Calle Gran Via 12 <Juan Perez>
Plaza Mayor 45 Madrid 28013 <Maria Gomez> +34-600-987-654
<Carlos Ruiz> +1-800-555-0199 Fifth Ave New York`;

const agenda1 = findInAgenda(agenda, "34-600-123-456");
console.log(agenda1);
// { name: "Juan Perez", address: "Calle Gran Via 12" }

const agenda2 = findInAgenda(agenda, "600-987");
console.log(agenda2);
// { name: "Maria Gomez", address: "Plaza Mayor 45 Madrid 28013" }

const agenda3 = findInAgenda(agenda, "111");
console.log(agenda3);
// null
// Explanation: No results

const agenda4 = findInAgenda(agenda, "1");
console.log(agenda4);
// null
// Explanation: Too many results

/* 
  Score: ★★★★★
*/
